hal/src/thumbv7em/icm.rs

```diff
@@ -0,0 +1,1612 @@
+    #[inline]
+    pub fn get_rsu_int(&self, bitmask: u8) -> u8 {
+        (self.interrupt_vector >> 20 & 0x0f) as u8 & bitmask
+    }
+
+    /// Region End bit Condition Detected interrupt status
+    #[inline]
+    pub fn get_rec_int(&self, bitmask: u8) -> u8 {
+        (self.interrupt_vector >> 16 & 0x0f) as u8 & bitmask
+    }
+
+    /// Region Wrap Condition detected interrupt status
+    #[inline]
+    pub fn get_rwc_int(&self, bitmask: u8) -> u8 {
+        (self.interrupt_vector >> 12 & 0x0f) as u8 & bitmask
+    }
+
+    /// Region Bus Error interrupt status
+    #[inline]
+    pub fn get_rbe_int(&self, bitmask: u8) -> u8 {
+        (self.interrupt_vector >> 8 & 0x0f) as u8 & bitmask
+    }
+
+    /// Region Digest Mis interrupt status
+    #[inline]
+    pub fn get_rdm_int(&self, bitmask: u8) -> u8 {
+        (self.interrupt_vector >> 4 & 0x0f) as u8 & bitmask
+    }
+
+    /// Region Hash Completed interrupt status
+    #[inline]
+    pub fn get_rhc_int(&self, bitmask: u8) -> u8 {
+        (self.interrupt_vector & 0x0f) as u8 & bitmask
+    }
+}
+impl Default for IcmInterrupt {
+    fn default() -> Self {
+        Self {
+            interrupt_vector: 0,
+        }
+    }
+}
+
+/// Struct useful for returning the interrupt status
+/// of the ICM. Provides methods for easy parsing of
+/// the region specific [`IcmRegionNum`]
+pub struct IcmRegionInterrupt<I: IcmRegionNum> {
+    region: PhantomData<I>,
+    interrupt_vector: u32,
+}
+
+impl<I: IcmRegionNum> IcmRegionInterrupt<I> {
+    /// Used to mask out the correct bit based on [`IcmRegionNum`]
+    #[inline]
+    fn mask(&self) -> u8 {
+        1 << I::NUM
+    }
+
+    /// Region Status Updated interrupt status
+    #[inline]
+    pub fn get_rsu_int(&self) -> bool {
+        matches!((self.interrupt_vector >> 20 & 0x0f) as u8 & self.mask(), 1)
+    }
+
+    /// Region End bit Condition Detected interrupt status
+    #[inline]
+    pub fn get_rec_int(&self) -> bool {
+        matches!((self.interrupt_vector >> 16 & 0x0f) as u8 & self.mask(), 1)
+    }
+
+    /// Region Wrap Condition detected interrupt status
+    #[inline]
+    pub fn get_rwc_int(&self) -> bool {
+        matches!((self.interrupt_vector >> 12 & 0x0f) as u8 & self.mask(), 1)
+    }
+
+    /// Region Bus Error interrupt status
+    #[inline]
+    pub fn get_rbe_int(&self) -> bool {
+        matches!((self.interrupt_vector >> 8 & 0x0f) as u8 & self.mask(), 1)
+    }
+
+    /// Region Digest Mismatches!( interrupt status
+    #[inline]
+    pub fn get_rdm_int(&self) -> bool {
+        matches!((self.interrupt_vector >> 4 & 0x0f) as u8 & self.mask(), 1)
+    }
+
+    /// Region Hash Completed interrupt status
+    #[inline]
+    pub fn get_rhc_int(&self) -> bool {
+        matches!((self.interrupt_vector & 0x0f) as u8 & self.mask(), 1)
+    }
+}
+
+/// IcmRegionToken provides access to region-specific
+/// settings like interrupts and status
+pub struct IcmRegionToken<I: IcmRegionNum> {
+    region: PhantomData<I>,
+}
```

Comment:
I wouldn't call this a `Token` type. I represents the public interface for an ICM region, so I would call it `IcmRegion` directly. I think you use that name further down, for the collection of four regions, but I would rename that to `IcmRegions`, with an `s`.

Separately, you might consider removing the `Icm` prefix from all your types. If users what to disambiguate, they can always use the `icm::` namespace. And removing the `Icm` prefix gives users the option to be more terse when context tells you that `Region` means `icm::Region`. It's up to you though.

Reply:
Adopted
* `IcmRegionToken` -> `Region`,
*  `IcmRegion` -> `Regions`